File: code/474. Ones and Zeroes.py
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[0 for j in range(n + 1)] for i in range(m + 1)]
        def count(s):
            c0, c1 = len(s), 0
            for ch in s:
                if ch == '1':
                    c0 -= 1
                    c1 += 1
            return c0, c1
        for s in strs:
            c0, c1 = count(s)
            for i in range(m, c0 - 1, -1):
                for j in range(n, c1 - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - c0][j - c1] + 1)
        return dp[m][n]
        
        

File: code/474. Ones and Zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        for (string& s : strs) {
            auto [count0, count1] = count(s);
            for (int i = m; i >= count0; i--) {
                for (int j = n; j >= count1; j--) {
                    dp[i][j] = max(dp[i][j], dp[i - count0][j - count1] + 1);
                }
            }
        }
        return dp[m][n];
    }
    pair<int, int> count(string s) {
        int count0 = s.length(), count1 = 0;
        for (char ch : s) {
            if (ch == '1') {
                ++count1;
                --count0;
            }
        }
        return make_pair(count0, count1);
    }
};